security/advancedtls/advancedtls_test.go

```diff
@@ -0,0 +1,482 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"reflect"
+	"strings"
+	"syscall"
+	"testing"
+
+	"google.golang.org/grpc/credentials"
+)
+
+func TestClientServerHandshake(t *testing.T) {
+	// ------------------Load Client Trust Cert and Server Peer Cert-------------------
+	clientTrustPool, err := readTrustCert("testdata/client_trust_cert_1.pem")
+	if err != nil {
+		t.Fatalf("Client is unable to load trust certs. Error: %v", err)
+	}
+	getRootCAsForClient := func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error) {
+		return clientTrustPool, nil
+	}
+	verifyFunc := func(serverName string, rawCerts [][]byte,
+		verifiedChains [][]*x509.Certificate) error {
+		if strings.HasPrefix(serverName, "127.0.0.1") {
+			return nil
+		}
+		return fmt.Errorf("custom verification function failed")
+
+	}
+	serverPeerCert, err := tls.LoadX509KeyPair("testdata/server_cert_1.pem",
+		"testdata/server_key_1.pem")
+	// ------------------Load Server Trust Cert and Client Peer Cert-------------------
+	serverTrustPool, err := readTrustCert("testdata/server_trust_cert_1.pem")
+	if err != nil {
+		t.Fatalf("Server is unable to load trust certs. Error: %v", err)
+	}
+	getRootCAsForServer := func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error) {
+		return serverTrustPool, nil
+	}
+	clientPeerCert, err := tls.LoadX509KeyPair("testdata/client_cert_1.pem",
+		"testdata/client_key_1.pem")
+	if err != nil {
+		t.Fatalf("Server is unable to parse certificates. Error: %v", err)
+	}
+	getRootCAsForServerBad := func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error) {
+		return nil, fmt.Errorf("bad root certificate reloading")
+	}
+	for _, test := range []struct {
+		desc                       string
+		clientCert                 []tls.Certificate
+		clientGetClientCert        func(*tls.CertificateRequestInfo) (*tls.Certificate, error)
+		clientRoot                 *x509.CertPool
+		clientGetRoot              func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
+		clientVerifyFunc           CustomVerificationFunc
+		clientExpectCreateError    bool
+		clientExpectHandshakeError bool
+		serverMutualTLS            bool
+		serverCert                 []tls.Certificate
+		serverGetCert              func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+		serverRoot                 *x509.CertPool
+		serverGetRoot              func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
+		serverExpectError          bool
+	}{
```

Comment:
We may want to test the cases where the root CA certs do not match the peer certificates. E.g., server provides a valid certificate, but server's CA cert is different from client Root cert.

Reply:
Added the following test cases:

- Client_reload_both_certs_verifyFunc_Server_reload_both_certs_mutualTLS
- Client_wrong_peer_cert_Server_reload_both_certs_mutualTLS
- Client_wrong_peer_cert_Server_reload_both_certs_mutualTLS
- Client_reload_both_certs_verifyFunc_Server_wrong_peer_cert
- Client_reload_both_certs_verifyFunc_Server_wrong_trust_cert